pandas/core/indexes/base.py

```diff
@@ -7017,7 +7017,11 @@ def diff(self, periods: int = 1) -> Self:
         Index([nan, 10.0, 10.0, 10.0, 10.0], dtype='float64')
 
         """
-        return self._constructor(self.to_series().diff(periods))
+        diff = self.to_series().diff(periods)
+        if diff.dtype != self.dtype:
+            # e.g. DTI.diff -> TDI
+            return Index(diff)
```

Comment:
let's just return Index(diff) unconditionally?

Reply:
I think we should use `self._constructor` in order to preserve subclasses as much as possible?

Reply:
are there any cases where it makes a difference?

Reply:
```python
class MyIndex(Index):
    def _constructor(self):
        return MyIndex

>>> MyIndex([1, 2]).diff(MyIndex([1, 3]))
```

Wouldn't it be a regression if we didn't return a `MyIndex` here?

Reply:
we dont support 3rd part Index subclasses

Reply:
It is noted in the docs that `Index` subclasses are not supported. @mroeschke, any objections to updating based on that?

Reply:
Ah okay I didn't know we didn't support Index subclasses. Fine to update to @jbrockmendel suggestion

Reply:
updated to return `Index(diff)` unconditionally